crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -369,8 +369,19 @@ impl<'w> EntityMut<'w> {
     ) {
         let old_archetype = &mut archetypes[old_archetype_id];
         let remove_result = old_archetype.swap_remove(old_location.archetype_row);
+        // if an entity was moved into this entity's archetype row, update its archetype row
         if let Some(swapped_entity) = remove_result.swapped_entity {
-            entities.set(swapped_entity.index(), old_location);
+            let swapped_location = entities.get(swapped_entity).unwrap();
+
+            entities.set(
+                swapped_entity.index(),
+                EntityLocation {
+                    archetype_id: swapped_location.archetype_id,
+                    archetype_row: old_location.archetype_row,
+                    table_id: swapped_location.table_id,
+                    table_row: swapped_location.table_row,
```

Comment:
```suggestion
                    archetype_row: old_location.archetype_row,
                    ..swapped_location
```
Alternatively, we should consider exposing a pub(crate) `get_mut` from `Entities` or just mutate swapped_location with the archetype row.

Reply:
I specifically chose to construct a whole new struct here so that if we change the fields we'll get a compile error rather than it continuing to compile

Reply:
That makes sense to me. Not sure if the compiler will optimize that down though. Probably worth investigating separately.

Not sure if this is worth leaving a comment so others don't try to make that change.